lib/Dialect/ArithExt/IR/ArithExtDialect.cpp

```diff
@@ -29,6 +30,28 @@ void ArithExtDialect::initialize() {
       >();
 }
 
+/// Ensures that the underlying integer type is wide enough for the coefficient
+template <typename OpType>
+LogicalResult verifyArithExtOpMod(OpType op) {
+  auto type = getElementTypeOrSelf(op.getResult().getType());
+  unsigned bitWidth = llvm::cast<IntegerType>(type).getWidth();
+  unsigned modWidth = (op.getModulus() - 1).getActiveBits();
+  if (modWidth > bitWidth)
+    return op.emitOpError()
+           << "underlying type's bitwidth must be at least as "
+           << "large as the modulus bitwidth, but got " << bitWidth
+           << " while modulus requires width " << modWidth << ".";
```

Comment:
Since it's not typed, we should be careful that a value of 2**32 - 1 is not interpreted as -1, as @asraa mentioned in another comment.

Can we add a verification to assert that, even when interpreting the modulus as signed, the value is always > 0, and then add a test for a modulus of 2**64 - 1 (64 is the default APInt bit width).

Reply:
Mh, very good point. I guess we shouldn't just hardcode to unsigned since we want to allow the underlying "container" type to be both signed (for people who prefer $[\floor{-q/2}, \floor{q/2})$) and unsigned (for $[0, q-1]$). For the modulus, though, e really only need unsignged now that I think about it... but I guess it's really the underlying type we care about so let's not mess with the attribute (which the parser makes signless 64-bit IntegerAttr, so if we were to change it, we'd have to always specify the type in the textual representation)

I'm not sure we can hard error on the > 0 check in all cases, though - it's a bit contrived, but someone might want their modulus to be 2^31-1 and use a 31-bit underlying type, which should pass the check here?
Maybe we should error if the underlying type is signed and the modulus isn't >0, warn if the underlying type is signless and the modulus would be <0 if interpreted as signed and not emit any warning/error if the underlying type is unsigned?

Reply:
> but someone might want their modulus to be 2^31-1 and use a 31-bit underlying type, which should pass the check here?

This is where I ran into issues with the remsi/remui before, since if the input is negative and the modulus is positive, you get the wrong answer either way (remsi gives 0 because the modulus is interpreted as -1, remui gives the wrong answer because the input is interpreted as a positive value mod 2**31).

I also think that this dialect is low-level enough that "preferences" about which coset representative is used are irrelevant. The lowering decides, and if we all agree to use $[0, q-1]$ then we can force it to be simpler that way.

I think specifying an explicit type in the textual IR will be safer, but I'm willing to give this a shot as-is.

Reply:
Oh, that's very annoying. :/  I guess supporting the  "q=2^31-1 in 31-bit underlying type" would require some kind of type conversion before we lower to `arith`?  That just seems way overkill, so I'll make it a hard error for both signless and signed then!

Reply:
I ended up having to make it "just" a warning for now because I realized (a) the ops only take `SignlessIntegerLike` so we couldn't test the "non-simple" add/sub cases with the hard error in place and (b) while I tried relaxing the type constraint to a quickly defined `IntegerLike`, I realized actually using unsigned types completely breaks the lowering to `arith`, which insists on `SignlessIntegerLike`. (Why though?)

Since we were already planning to discuss data represetations/signedness stuff in tomorrow's meeting, I'll just add this to the pile :)

Reply:
Signless is required, I knew that. I just meant we should only be constraining the modulus, and interpreting the input values as part of the semantics.